moveit_planners/ompl/ompl_interface/src/detail/ompl_constraints.cpp

```diff
@@ -323,6 +323,166 @@ Eigen::MatrixXd OrientationConstraint::calcErrorJacobian(const Eigen::Ref<const
+      box_pose.position.y = 0.5 * (pose.translation().y() + next_pose.translation().y());
+      box_pose.position.z = 0.5 * (pose.translation().z() + next_pose.translation().z());
+      box_pose.orientation.w = 1.0;  // TODO: Orient the box
+      box_constraint.constraint_region.primitive_poses.push_back(box_pose);
+      box_constraint.weight = 1.0;
+
+      moveit_msgs::msg::Constraints constr;
+      constr.position_constraints.push_back(box_constraint);
+
+      pos_con->init(constr);
+      box_constraints_.emplace_back(pos_con);
+      // std::cout << "Made box constraint index " << i << std::endl;
+    }
+  }
+  else
+  {
+    RCLCPP_ERROR_STREAM(LOGGER, "Path constraint must specify at least 2 points!");
+    return;
+  }
+
+  // Extract the interpolation distance
+  if (path_constraint.interp_distance > 0.0)
+  {
+    interp_distance_ = path_constraint.interp_distance;
+  }
+  else
+  {
+    RCLCPP_WARN_STREAM(LOGGER, "Cannot use zero interpolation distance. Defaulting to 0.01.");
+    interp_distance_ = 0.01;
+  }
+
+  // Interpolate along waypoints and insert into pose_nn_;
+  for (size_t i = 0; i < num_pts - 1; i++)
+  {
+    // Insert waypoints
+    pose_nn_.add(path_[i]);
+
+    // TODO: Consider arc length between the poses to determine the number of interpolation steps
+    // Linearly determine the number of steps
+    const auto path_length = (path_[i].tform_.translation() - path_[i + 1].tform_.translation()).norm();
+    const auto num_steps = static_cast<unsigned int>(path_length / interp_distance_);
+
+    // Insert interpolated poses
+    auto delta = interp_distance_;
+    for (size_t j = 0; j < num_steps - 1; j++)
+    {
+      const auto pose = interpolatePose(path_[i].tform_, path_[i + 1].tform_, delta);
+      pose_nn_.add(EigenIsometry3dWrapper(pose, i));
+      delta += interp_distance_;
+    }
+  }
+
+  // Insert last waypoint
+  pose_nn_.add(path_[num_pts - 1]);
+}
+
+void ToolPathConstraint::function(const Eigen::Ref<const Eigen::VectorXd>& joint_values,
+                                  Eigen::Ref<Eigen::VectorXd> out) const
+{
+  const Eigen::Isometry3d eef_pose = forwardKinematics(joint_values);
+
+  const auto eef_nearest = pose_nn_.nearest(EigenIsometry3dWrapper(eef_pose));
+  // std::cout << "Function, nearest: " << eef_nearest.path_idx_ << std::endl;
+
+  const auto bc = box_constraints_.at(eef_nearest.path_idx_);
+  bc->function(joint_values, out);
+}
+
+void ToolPathConstraint::jacobian(const Eigen::Ref<const Eigen::VectorXd>& joint_values,
+                                  Eigen::Ref<Eigen::MatrixXd> out) const
+{
+  const Eigen::Isometry3d eef_pose = forwardKinematics(joint_values);
+  const auto eef_nearest = pose_nn_.nearest(EigenIsometry3dWrapper(eef_pose));
+  // std::cout << "Jacobian, nearest: " << eef_nearest.path_idx_ << std::endl;
+
+  const auto bc = box_constraints_.at(eef_nearest.path_idx_);
+  bc->jacobian(joint_values, out);
+}
```

Comment:
These are the "magic" functions that map the EEF pose to the nearest box constraint for that path segment and compute the standard function/Jacobian for that box constraints.

I think this is close, except that I think the state sampler needs to be configured.